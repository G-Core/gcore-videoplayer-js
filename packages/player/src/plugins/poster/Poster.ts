//Copyright 2014 Globo.com Player authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

import {
  Events,
  Playback,
  PlayerError,
  UIContainerPlugin,
  template,
  $,
  Container,
} from '@clappr/core'
import { trace } from '@gcorevideo/utils'

import { CLAPPR_VERSION } from '../../build.js'
import type { ZeptoResult } from '../../utils/types.js'

import '../../../assets/poster/poster.scss'
import posterHTML from '../../../assets/poster/poster.ejs'
import playIcon from '../../../assets/icons/new/play.svg'
import { PlaybackError } from '../../playback.types.js'

const T = 'plugins.poster_custom'

/**
 * Displays a poster image in the background and a big play button on top when playback is stopped
 * @beta
 * @remarks
 * When the playback is stopped, media control UI is disabled.
 *
 * Configuration options:
 *
 * - `poster.custom` - custom CSS background
 *
 * - `poster.showForNoOp` - whether to show the poster when the playback is not started
 *
 * - `poster.url` - the URL of the poster image
 *
 * - `poster.showOnVideoEnd` - whether to show the poster when the playback is ended
 *
 * @example
 * ```ts
 * new Player({
 *  ...
 *  poster: {
 *    showForNoOp: true,
 *    url: 'https://via.placeholder.com/150.png',
 *  }
 * })
 * ```
 */
export class Poster extends UIContainerPlugin {
  private hasFatalError = false

  private hasStartedPlaying = false

  private playRequested = false

  private $playButton: ZeptoResult | null = null

  private $playWrapper: ZeptoResult | null = null

  /**
   * @internal
   */
  get name() {
    return 'poster_custom'
  }

  /**
   * @internal
   */
  get supportedVersion() {
    return { min: CLAPPR_VERSION }
  }

  private static readonly template = template(posterHTML)

  private get shouldRender() {
    if (!this.enabled || this.options.reloading) {
      return false
    }
    const showForNoOp = !!this.options.poster?.showForNoOp
    return (
      this.container.playback.name !== 'html_img' &&
      (this.container.playback.getPlaybackType() !== Playback.NO_OP ||
        showForNoOp)
    )
  }

  /**
   * @internal
   */
  override get attributes() {
    return {
      class: 'player-poster',
      'data-poster': '',
    }
  }

  /**
   * @internal
   */
  override get events() {
    return {
      click: 'clicked',
    }
  }

  private get showOnVideoEnd() {
    return this.options.poster?.showOnVideoEnd !== false
  }

  /**
   * @internal
   */
  override bindEvents() {
    this.listenTo(this.container, Events.CONTAINER_STOP, this.onStop)
    this.listenTo(this.container, Events.CONTAINER_PLAY, this.onPlay)
    this.listenTo(this.container, Events.CONTAINER_STATE_BUFFERING, this.update)
    this.listenTo(
      this.container,
      Events.CONTAINER_STATE_BUFFERFULL,
      this.update,
    )
    this.listenTo(this.container, Events.CONTAINER_OPTIONS_CHANGE, this.render)
    this.listenTo(this.container, Events.CONTAINER_ERROR, this.onError)
    this.showOnVideoEnd &&
      this.listenTo(this.container, Events.CONTAINER_ENDED, this.onStop)
    this.listenTo(this.container, Events.CONTAINER_READY, this.render)
    this.listenTo(this.container, Events.PLAYBACK_PLAY_INTENT, this.onPlayIntent)
  }

  /**
   * Reenables earlier disabled plugin
   */
  override enable() {
    super.enable()
    this.hasStartedPlaying = this.container.playback.isPlaying()
    this.update()
  }

  /**
   * Disables the plugin, unmounting it from the DOM
   */
  override disable() {
    trace(`${T} disable`)
    this.hasStartedPlaying = false
    this.playRequested = false
    super.disable()
  }

  private onError(error: PlaybackError) {
    trace(`${T} onError`, {
      error,
      enabled: this.enabled,
    })
    this.hasFatalError = error.level === PlayerError.Levels.FATAL

    if (this.hasFatalError) {
      this.hasStartedPlaying = false
      if (!this.playRequested) {
        this.showPlayButton()
      }
    }
  }

  private onPlay() {
    trace(`${T} onPlay`)
    this.hasStartedPlaying = true
    this.playRequested = false
    this.update()
  }

  private onPlayIntent() {
    trace(`${T} onPlayIntent`)
    this.playRequested = true
    this.update()
  }

  private onStop() {
    trace(`${T} onStop`, {
      enabled: this.enabled,
    })
    this.hasStartedPlaying = false
    this.playRequested = false
    this.update()
  }

  private updatePlayButton(show: boolean) {
    trace(`${T} updatePlayButton`, {
      show,
      chromeless: this.options.chromeless,
      allowUserInteraction: this.options.allowUserInteraction,
    })
    if (
      show &&
      (!this.options.chromeless || this.options.allowUserInteraction)
    ) {
      this.showPlayButton()
    } else {
      this.hidePlayButton()
    }
  }

  private showPlayButton() {
    if (this.options.disableMediaControl) {
      return
    }
    if (this.hasFatalError && !this.options.disableErrorScreen) {
      return
    }

    this.$playButton?.show()
    this.$el.addClass('clickable')
    this.container.$el.addClass('container-with-poster-clickable')
  }

  private hidePlayButton() {
    this.$playButton.hide()
    this.$el.removeClass('clickable')
  }

  private clicked() {
    trace(`${T} clicked`, {
      hasStartedPlaying: this.hasStartedPlaying,
      chromeless: this.options.chromeless,
      allowUserInteraction: this.options.allowUserInteraction,
    })
    // Let "click_to_pause" plugin handle click event if media has started playing
    if (!this.hasStartedPlaying) {
      if (!this.options.chromeless || this.options.allowUserInteraction) {
        this.playRequested = true
        this.update()
        this.container.playback.consent()
        this.container.playback.play()
      }
    } else {
      this.container.trigger('container:start')
    }

    return false
  }

  private shouldHideOnPlay() {
    // Audio broadcasts should keep the poster up; video should hide poster while playing.
    return !this.container.playback.isAudioOnly
  }

  private update() {
    trace(`${T} update`, {
      shouldRender: this.shouldRender,
    })
    if (!this.shouldRender) {
      return
    }

    const showPlayButton =
      !this.playRequested &&
      !this.hasStartedPlaying &&
      !this.container.buffering

    this.updatePlayButton(showPlayButton)
    this.updatePoster()
  }

  private updatePoster() {
    trace(`${T} updatePoster`, {
      hasStartedPlaying: this.hasStartedPlaying,
    })
    if (!this.hasStartedPlaying) {
      this.showPoster()
    } else {
      this.hidePoster()
    }
  }

  private showPoster() {
    this.container.disableMediaControl()
    this.$el.show()
  }

  private hidePoster() {
    trace(`${T} hidePoster`, {
      shouldHideOnPlay: this.shouldHideOnPlay(),
    })
    if (!this.options.disableMediaControl) {
      this.container.enableMediaControl()
    }
    if (this.shouldHideOnPlay()) {
      this.$el.hide()
    }
  }

  /**
   * @internal
   */
  override render() {
    if (!this.shouldRender) {
      return this
    }

    this.$el.html(Poster.template())

    const isRegularPoster =
      this.options.poster && this.options.poster.custom === undefined

    if (isRegularPoster) {
      const posterUrl = this.options.poster.url || this.options.poster

      this.$el.css({ 'background-image': 'url(' + posterUrl + ')' })
    } else if (this.options.poster) {
      this.$el.css({ background: this.options.poster.custom })
    }

    this.container.$el.removeClass('container-with-poster-clickable')
    this.container.$el.append(this.el)
    this.$playWrapper = this.$el.find('.play-wrapper')
    this.$playWrapper.addClass('control-need-disable')
    this.$playButton = $(
      "<div class='circle-poster gcore-skin-button-color gcore-skin-border-color'></div>",
    )
    this.$playWrapper.append(this.$playButton)
    this.$playButton.append(playIcon)

    if (this.options.autoPlay) {
      this.$playButton.hide()
    }
    this.$playButton.addClass('poster-icon')
    this.$playButton.attr('data-poster', '')

    this.update()

    return this
  }

  /**
   * @internal
   */
  override destroy() {
    this.container.$el.removeClass('container-with-poster-clickable')
    return this
  }
}
