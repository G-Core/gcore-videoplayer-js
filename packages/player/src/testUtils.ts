import { $, UICorePlugin } from '@clappr/core'
import Events from 'eventemitter3'
import { vi } from 'vitest'

export function createMockCore(
  options: Record<string, unknown> = {},
  container: any = createMockContainer(options),
) {
  const el = document.createElement('div')
  const emitter = new Events()
  return Object.assign(emitter, {
    el,
    $el: $(el),
    activePlayback: container.playback,
    activeContainer: container,
    i18n: {
      t: vi.fn().mockImplementation((key: string) => key),
    },
    options: {
      ...options,
    },
    configure: vi.fn(),
    getPlaybackType: vi.fn(),
    getPlugin: vi.fn(),
    load: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockPlugin() {
  return Object.assign(new Events(), {
    enable: vi.fn(),
    disable: vi.fn(),
  })
}

export function createSpinnerPlugin() {
  return Object.assign(createMockPlugin(), {
    show: vi.fn(),
    hide: vi.fn(),
  })
}

export function createMockPlayback(name = 'mock') {
  const emitter = new Events()
  return Object.assign(emitter, {
    name,
    currentLevel: -1,
    el: document.createElement('video'),
    dvrEnabled: false,
    dvrInUse: false,
    isAudioOnly: false,
    levels: [],
    consent: vi.fn(),
    play: vi.fn(),
    pause: vi.fn(),
    stop: vi.fn(),
    destroy: vi.fn(),
    seek: vi.fn(),
    seekPercentage: vi.fn(),
    getDuration: vi.fn().mockImplementation(() => 100),
    enterPiP: vi.fn(),
    exitPiP: vi.fn(),
    getPlaybackType: vi.fn(),
    getStartTimeOffset: vi.fn().mockImplementation(() => 0),
    getCurrentTime: vi.fn().mockImplementation(() => 0),
    isHighDefinitionInUse: vi.fn().mockImplementation(() => false),
    mute: vi.fn(),
    unmute: vi.fn(),
    volume: vi.fn(),
    configure: vi.fn(),
    attemptAutoPlay: vi.fn().mockImplementation(() => true),
    canAutoPlay: vi.fn().mockImplementation(() => true),
    onResize: vi.fn().mockImplementation(() => true),
    setPlaybackRate: vi.fn(),
    switchAudioTrack: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockContainer(
  options: Record<string, unknown> = {},
  playback: any = createMockPlayback(),
) {
  const el = playback.el
  const emitter = new Events()
  return Object.assign(emitter, {
    el,
    playback,
    options: {
      ...options,
    },
    $el: $(el),
    disableMediaControl: vi.fn(),
    enableMediaControl: vi.fn(),
    enterPiP: vi.fn(),
    exitPiP: vi.fn(),
    getDuration: vi.fn().mockReturnValue(0),
    getPlugin: vi.fn(),
    getPlaybackType: vi.fn(),
    getStartTimeOffset: vi.fn().mockReturnValue(0),
    isDvrInUse: vi.fn().mockReturnValue(false),
    isDvrEnabled: vi.fn().mockReturnValue(false),
    isHighDefinitionInUse: vi.fn().mockReturnValue(false),
    isPlaying: vi.fn().mockReturnValue(false),
    play: vi.fn(),
    seek: vi.fn(),
    seekPercentage: vi.fn(),
    switchAudioTrack: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockMediaControl(core: any) {
  const mediaControl = new UICorePlugin(core)
  // TODO <div class="media-control-layer">
  mediaControl.$el.html(
    `<div class="media-control-left-panel" data-media-control></div>
    <div class="media-control-right-panel" data-media-control></div>
    <div class="media-control-center-panel" data-media-control></div>`,
  )
  // @ts-ignore
  mediaControl.mount = vi.fn()
  // @ts-ignore
  mediaControl.container = core.activeContainer
  // @ts-ignore
  mediaControl.toggleElement = vi.fn()
  vi.spyOn(mediaControl, 'trigger')
  core.$el.append(mediaControl.$el)
  return mediaControl
}

export function createMockBottomGear(core: any) {
  const plugin: any = new UICorePlugin(core)
  plugin.getItem = vi.fn()
  plugin.addItem = vi.fn().mockImplementation((name: string, $el: any) => {
    const existing = plugin.$el.find(`[data-${name}]`)
    if (existing.length) {
      return existing
    }
    return $('<li></li>')
      .attr(`data-${name}`, '')
      .append($el)
      .appendTo(plugin.$el)
  })
  plugin.refresh = vi.fn()
  return plugin
}
