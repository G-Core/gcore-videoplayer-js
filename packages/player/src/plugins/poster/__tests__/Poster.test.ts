import { Poster } from '../Poster'
import { describe, it, expect, beforeEach } from 'vitest'
import { createMockCore } from '../../../testUtils'
import { Events, PlayerError } from '@clappr/core'

describe('Poster', () => {
  let core: any
  let poster: Poster
  describe('basically', () => {
    beforeEach(() => {
      core = createMockCore({
        poster: {
          url: 'https://via.placeholder.com/150.png',
        },
      })
      poster = new Poster(core.activeContainer)
      core.activeContainer.trigger(Events.CONTAINER_READY)
    })
    it('should render', () => {
      expect(poster.el.innerHTML).toMatchSnapshot()
    })
    describe('when clicked', () => {
      beforeEach(() => {
        poster.el.click()
      })
      it('should start playback', () => {
        expect(core.activeContainer.play).toHaveBeenCalled()
        expect(core.activeContainer.playback.consent).not.toHaveBeenCalled()
      })
      it('should hide button', () => {
        expect(poster.$el.find('#poster-play')[0].style.display).toBe('none')
      })
      it('should remove clickable class', () => {
        expect(poster.el.classList.contains('clickable')).toBe(false)
      })
    })
    describe('when playback is triggered', () => {
      beforeEach(() => {
        core.activeContainer.playback.trigger(Events.PLAYBACK_PLAY_INTENT)
      })
      it('should hide button', () => {
        expect(poster.$el.find('#poster-play')[0].style.display).toBe('none')
      })
    })
    describe('when playback is about to start', () => {
      describe.each([
        [
          Events.CONTAINER_STATE_BUFFERING, 
        ], [
          Events.CONTAINER_STATE_BUFFERFULL,
        ]
      ])("event %s", (event) => {
        beforeEach(() => {
          core.activeContainer.buffering = true
          core.activeContainer.trigger(event)
        })
        it('should hide button', () => {
          expect(poster.$el.find('#poster-play')[0].style.display).toBe('none')
        })
      })
    })
    describe('when playback is started', () => {
      beforeEach(() => {
        core.activeContainer.trigger(Events.CONTAINER_PLAY)
            core.activeContainer.playback.trigger(Events.PLAYBACK_PLAY)
      })
      it('should hide poster', () => {
        expect(poster.el.style.display).toBe('none')
      })
    })
    describe('when playback is stopped', () => {
      beforeEach(() => {
        core.activeContainer.trigger(Events.CONTAINER_PLAY)
        core.activeContainer.playback.trigger(Events.PLAYBACK_PLAY)
        core.activeContainer.trigger(Events.CONTAINER_STOP)
      })
      it('should show poster', () => {
        expect(poster.el.style.display).not.toBe('none')
      })
      it('should show button', () => {
        expect(poster.$el.find('#poster-play')[0].style.display).not.toBe('none')
      })
      it('should add clickable class', () => {
        expect(poster.el.classList.contains('clickable')).toBe(true)
      })
    })
  })
  describe('when autoplay is configured', () => {
    beforeEach(() => {
      core = createMockCore({
        autoPlay: true,
      })
      poster = new Poster(core.activeContainer)
      core.activeContainer.trigger(Events.CONTAINER_READY)
    })
    it('should hide button initially', () => {
      expect(poster.$el.find('#poster-play')[0].style.display).toBe('none')
    })
  })
  describe('when error occurs', () => {
    beforeEach(() => {
      core = createMockCore({
        autoPlay: true,
      })
      poster = new Poster(core.activeContainer)
      core.activeContainer.trigger(Events.CONTAINER_READY)
      core.activeContainer.playback.trigger(Events.PLAYBACK_PLAY_INTENT)
      core.activeContainer.trigger(Events.CONTAINER_ERROR, {
        level: PlayerError.Levels.FATAL,
      })
    })
    it('should show poster', () => {
      expect(poster.el.style.display).not.toBe('none')
    })
    it('should hide button', () => {
      expect(poster.$el.find('#poster-play')[0].style.display).toBe('none')
    })
  })
})