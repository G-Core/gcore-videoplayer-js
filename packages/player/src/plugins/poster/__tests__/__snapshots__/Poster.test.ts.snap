// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Poster > basically > should render 1`] = `
"<div class="play-wrapper" id="gplayer-poster">
    <div class="circle-poster gcore-skin-button-color gcore-skin-border-color" id="poster-play">/assets/icons/new/play.svg</div>
</div>
"
`;
